src/nnet3/nnet-simple-component.h

```diff
@@ -873,6 +875,83 @@ class PerElementScaleComponent: public UpdatableComponent {
 };
 
 
+// PerElementOffsetComponent offsets each dimension of its input with a separate
+// trainable bias; it's like an affine component with fixed weight matrix which is always equal to I.
+class PerElementOffsetComponent: public UpdatableComponent {
+ public:
+  virtual int32 InputDim() const { return offsets_.Dim(); }
+  virtual int32 OutputDim() const { return offsets_.Dim(); }
+
+  virtual std::string Info() const;
+  virtual void InitFromConfig(ConfigLine *cfl);
+
+  PerElementOffsetComponent() { } // use Init to really initialize.
+  virtual std::string Type() const { return "PerElementOffsetComponent"; }
+  virtual int32 Properties() const {
```

Comment:
Regarding kPropagateAdds and kBackpropAdds: it's a question of what is most
convenient for the propagate and backprop code, i.e. what will avoid extra
operations.  For instance, in the affine component we use kPropagateAdds
and kBackpropAdds because AddMatMat adds to its output (and a zero
coefficient doesn't stop this, because it propagates infinities).
Dan

On Sun, Nov 1, 2015 at 5:40 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.h
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43588951>:
>
> > @@ -873,6 +875,83 @@ class PerElementScaleComponent: public UpdatableComponent {
> >  };
> >
> >
> > +// PerElementOffsetComponent offsets each dimension of its input with a separate
> > +// trainable bias; it's like an affine component with fixed weight matrix which is always equal to I.
> > +class PerElementOffsetComponent: public UpdatableComponent {
> > + public:
> > +  virtual int32 InputDim() const { return offsets_.Dim(); }
> > +  virtual int32 OutputDim() const { return offsets_.Dim(); }
> > +
> > +  virtual std::string Info() const;
> > +  virtual void InitFromConfig(ConfigLine *cfl);
> > +
> > +  PerElementOffsetComponent() { } // use Init to really initialize.
> > +  virtual std::string Type() const { return "PerElementOffsetComponent"; }
> > +  virtual int32 Properties() const {
>
> You can also add kBackpropInPlace. This component is not
> kLinearInParameters.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43588951>.
>